### PR TITLE
Enable docker image feature with xpk

### PR DESCRIPTION
**Pull request by RissyRan**
New [PR](https://github.com/GoogleCloudPlatform/ml-auto-solutions/pull/40).

# Description

Enable docker image feature with xpk:
* Add this as an GKE example, and plan to add more examples for quick start
* Add `TpuGkeTask` to execute run_model and post_process task groups 
* Add cluster configs under `configs/cluster` folder, so Airflow has the access to the created cluster with TPUs. It seems `certificate-authority-data` is safe to share based on this stackoverflow [thread](https://stackoverflow.com/questions/71444145/should-the-k8s-cluster-certificate-authority-be-kept-secret).
* Update `TpuTask` to `TpuGceTask` to distinguish `TpuGkeTask`, and similar for test config

A few TODO once b/311073979 is approved:
* Update GCP and cluster configs, and host clusters in cloud-ml-auto-solutions project
* Will have a separate PR to document how to provision cluster using XPK. It is quite straightforward with 1-2 commands

# Tests

Please describe the tests that you ran on Cloud VM to verify changes.

**Instruction and/or command lines to reproduce your tests:** ...
* Upload codes to GCS bucket and test

**List links for your tests (use go/shortn-gen for any internal link):** ...
* Airflow test: [link](https://screenshot.googleplex.com/4uVjRBGjNY5B6Q5)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run one-shot tests and provided workload links above if applicable. 
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: dags/benchmark/pytorchxla_torchbench.py [hunk 1]
```diff
@@ -31,9 +31,9 @@
 SCHEDULED_TIME = "0 17 * * *" if composer_env.is_prod_env() else None
 
 with models.DAG(
-    dag_id="pytorchxla-torchbench",
-    schedule=SCHEDULED_TIME,
-    tags=["pytorchxla", "nightly", "torchbench"],
+    dag_id="pytorch_nightly_torchbench",
+    schedule=None,
```
Comment (will-cromar) [inline dags/benchmark/pytorchxla_torchbench.py:35]: Is this intentional? Do we want to disable this test's schedule?

File: implementations/utils/xpk.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    accelerator_type: str,
+    run_cmds: str,
+    task_owner: str,
+    num_slices: int = 1,
+) -> KubernetesPodOperator:
+  """Run workload through xpk tool.
+
+  The reason to use KubernetesPodOperator instead of BashOperator is that
+  xpk must run with Python 3.10 or greater; however, the latest version in
+  Composer is Python 3.8, and it's non-trivial to upgrade it as the  Composer
```
Comment (will-cromar) [inline implementations/utils/xpk.py:48]: :facepalm: I wish they would update this

Airflow actually provides a neat tool for this situation. You can run an `@task` directly in a Docker container with a different Python environment: https://airflow.apache.org/docs/apache-airflow/stable/tutorial/taskflow.html#dependency-separation-using-docker-operator

This gives us at least one less level of indirection: a docker container on the current host, rather than a docker container on another host. There's also a corresponding `@task.kubernetes` in case we do need a cluster!

File: implementations/utils/metric.py [hunk 3]
```diff
@@ -397,8 +396,38 @@ def is_valid_entry() -> bool:
...
+  """Get job status for the GKE run.
+
+  FAILED - if any failure occurs in run_model
+  SUCCESS - end-to-end model tests are successful in run_model
+  """
+  context = get_current_context()
+  execution_date = context["dag_run"].logical_date
+  current_dag = context["dag"]
+
+  workload_completion = current_dag.get_task(
```
Comment (will-cromar) [inline implementations/utils/metric.py:409]: This DAG introspection looks kind of sketchy to me. Could the status here be plumbed through XCOM instead?

File: implementations/utils/xpk.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      owner=task_owner,
+  )
+
+
+@task.sensor(poke_interval=60, timeout=600, mode="reschedule")
+def wait_for_workload_completion(workload_id: str, cluster_config: str) -> bool:
+  """Check the workload status."""
+
+  # Load the config for the cluster with TPUs in the pool
+  config.load_kube_config(
```
Comment (will-cromar) [inline implementations/utils/xpk.py:84]: If it's not too verbose, do you think it would be clearer to `import kubernetes` and write this as `kubernetes.config.etc`? Otherwise, the `config` and `client` modules are a little ambiguous

File: configs/example/gke_example_config.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    time_out_in_min: int,
+) -> task.TpuGkeTask:
+  # TODO(ranran): update the project once quota is approved (b/311073979).
+  job_gcp_config = gcp_config.GCPConfig(
+      project_name="tpu-prod-env-one-vm",
+      zone=tpu_zone,
+      dataset_name=metric_config.DatasetOption.XLML_DATASET,
+  )
+
+  run_model_cmds = (
```
Comment (will-cromar) [inline configs/example/gke_example_config.py:37]: If this test is actually going to be used, should it be in the `example` directory? If it is just an example, I would go even simpler and just print `jax.device_count()` to show the configuration is correct.

File: configs/cluster/v5e_cluster_config [hunk 6]
```diff
@@ -0,0 +1,26 @@
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply configs/cluster/v5e_cluster_config:1]: Okay, way back in ancient history we used to do something similar in the old implementation of the metrics handler. This is how we built the kube config programatically: https://github.com/GoogleCloudPlatform/ml-testing-accelerators/blob/3bd5720e667207fc6a49e8f692110fd812f4fc5b/metrics_handler/job_status_handler.py#L59-L94

File: apis/task.py [hunk 7]
```diff
@@ -41,15 +41,15 @@ def run() -> DAGNode:
 
 
 @dataclasses.dataclass
-class TpuTask(BaseTask):
-  """This is a class to set up tasks for TPU.
+class TpuGceTask(BaseTask):
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply apis/task.py:44]: Yes, we could. If that's the case, I'd like to update `TpuGkeTask` to be `TpuXpkTask` to be specific as well.

File: configs/vm_resource.py [hunk 8]
```diff
@@ -32,3 +32,20 @@ class RuntimeVersion(enum.Enum):
...
+
+# TODO(ranran): update the cluster name once quota is approved (b/311073979).
+class ClusterName(enum.Enum):
+  V4_CLUSTER = ""
+  V5E_CLUSTER = "ran-xpk-test-zone"
+
+
+# TODO(ranran): update the cluster name once quota is approved (b/311073979).
+class ClusterConfig(enum.Enum):
+  V4_CONFIG = "v4_cluster_config"
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply configs/vm_resource.py:45]: Thanks! My understanding the way xpk works is to use all of TPUs in a cluster, and it cannot use partial of them. i.e. we have a cluster in region us-central-2, and have TPU v4 and v5e in different pool ([example](https://screenshot.googleplex.com/BFNC7bpjyBFPzFP)). But I will need to double check with Victor on this.

File: dags/benchmark/pytorchxla_torchbench.py (hunk 1 above)
Comment (RissyRan) [reply dags/benchmark/pytorchxla_torchbench.py:35]: Oh.. I think this is introduced due to the merge a while ago. The head has been changed already, I will update in the next commit.

File: implementations/utils/xpk.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  """
+
+  cmds = (
+      "set -x",
+      f"gcloud config set project {project_id}",
+      f"gcloud config set compute/zone {zone}",
+      "git clone -b xpk-namespace https://github.com/google/xpk.git /tmp/xpk",
+      "cd /tmp/xpk",
+      (
+          "python3 xpk.py workload create"
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply implementations/utils/xpk.py:60]: I don't. All examples are provided for bash script here (https://github.com/google/xpk), but I will double check with Victor on this.

File: implementations/utils/metric.py (hunk 3 above)
Comment (RissyRan) [reply implementations/utils/metric.py:409]: I did not find a good way to get the status when running `run_workload` via bash script. There is a bug for xpk to wait till completion (b/306688270). I tried command in that bug to get status of workload, but unfortunately, we will need extra efforts to clean up output (output including extra logs & tables).

File: implementations/utils/xpk.py (hunk 4 above)
Comment (RissyRan) [reply implementations/utils/xpk.py:84]: Sounds good!

File: configs/example/gke_example_config.py (hunk 5 above)
Comment (RissyRan) [reply configs/example/gke_example_config.py:37]: This is a good point. Yes, currently I just want to show an example. Since we are testing models, I was thinking probably it's better to have a hello world E2E example. Do you feel this is a complex one?

File: implementations/utils/metric.py (hunk 3 above)
Comment (RissyRan) [reply implementations/utils/metric.py:409]: One example for status: https://screenshot.googleplex.com/BpDxsaUX6nh6BJV We will need to parse both status and status message to see if this job is failed or successful.

File: apis/test_config.py [hunk 10]
```diff
@@ -78,7 +79,7 @@ class Tpu(Accelerator):
 
   version: str
   cores: int
-  runtime_version: str
+  runtime_version: str = vm_resource.RuntimeVersion.TPU_UBUNTU2204_BASE.value
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply apis/test_config.py:82]: Good call! Updated it to `runtime_version: Optional[str] = None`

File: implementations/utils/xpk.py (hunk 4 above)
Comment (RissyRan) [reply implementations/utils/xpk.py:84]: Update to something like `from kubernetes import client as kubernetes_client, config as kubernetes_config`

File: implementations/utils/xpk.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply implementations/utils/xpk.py:60]: No, unfortunately, there is no lib plan on xpk.

File: implementations/utils/xpk.py (hunk 2 above)
Comment (RissyRan) [reply implementations/utils/xpk.py:48]: Discussed offline. It seems SubprocessHook is not compatible with @task.docker and @task.kubernetes. Not sure if this is Composer specific.

File: implementations/utils/metric.py (hunk 3 above)
Comment (will-cromar) [reply implementations/utils/metric.py:409]: Got it. Let's revisit this if/when XPK adds a Python API and see if there's a better way to plumb the status.
